src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+                if(remainingSlicesInLayer > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * (remainingSlicesInLayer - 1u);
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
+            else if (currentBlockInRow > 0)
+            {
+                memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[0] * remainingBlocksInRow;
+                if(remainingRowsInSlice > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[1] * (remainingRowsInSlice - 1u);
+                if(remainingSlicesInLayer > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[2] * (remainingSlicesInLayer - 1u);
+                if(remainingLayersInRegion > 1u)
+                    memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * (remainingLayersInRegion - 1u);
+            }
+        }
+        else
+        {
+            memoryNeededForRemainingRegions += imageExtentBlockStridesInBytes[3] * region.imageSubresource.layerCount; // = blockByteSize * imageExtentInBlocks.x * imageExtentInBlocks.y * imageExtentInBlocks.z * region.imageSubresource.layerCount
+        }
+    }
+    return memoryNeededForRemainingRegions;
+}
+
+bool ImageRegionIterator::advanceAndCopyToStagingBuffer(asset::IImage::SBufferCopy& regionToCopyNext, size_t& availableMemory, size_t& stagingBufferOffset, void* stagingBufferPointer)
+{
+    if(isFinished())
+        return false;
+        
+    auto addToCurrentUploadBufferOffset = [&](uint32_t size) -> bool 
+    {
+        const auto initialOffset = stagingBufferOffset;
+        stagingBufferOffset += size;
+        stagingBufferOffset = core::alignUp(stagingBufferOffset, bufferOffsetAlignment);
+        const auto consumedMemory = stagingBufferOffset - initialOffset;
+        if(consumedMemory <= availableMemory)
+        {
+            availableMemory -= consumedMemory;
+            return true;
+        }
+        else
+        {
+            assert(false);
+            return false;
+        }
+    };
+
+    if(!addToCurrentUploadBufferOffset(0u)) // in order to fix initial alignment to bufferOffsetAlignment
+    {
+        return false;
+    }
+
+    // ! Current Region that may break down into smaller regions (the first smaller region is nextRegionToCopy)
+    const asset::IImage::SBufferCopy & mainRegion = regions[currentRegion];
+        
+    asset::TexelBlockInfo srcImageTexelBlockInfo(srcImageFormat);
+    asset::TexelBlockInfo dstImageTexelBlockInfo(dstImageFormat);
+        
+    const core::vector4du32_SIMD srcBufferByteStrides = srcImageTexelBlockInfo.convert3DBlockStridesTo1DByteStrides(mainRegion.getBlockStrides(srcImageTexelBlockInfo));
```

Comment:
don't ever use this, each `SBufferCopy` can set its own row pitches and slice heights (custom alignments), use following family of functions
https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/nbl/asset/IImage.h#L173-L196